File: src/test/java/com/github/nsc/de/shake/lexer/TokenPosition.java
package com.github.nsc.de.shake.lexer;

import com.github.nsc.de.shake.lexer.characterinputstream.CharacterInputStream;
import com.github.nsc.de.shake.lexer.characterinputstream.SourceCharacterInputStream;
import com.github.nsc.de.shake.lexer.characterinputstream.position.Position;
import com.github.nsc.de.shake.lexer.token.TokenInputStream;
import com.github.nsc.de.shake.lexer.token.TokenType;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class TokenPosition {

    @Test
    public void testGetBasicPosition() {

        CharacterInputStream chars = new SourceCharacterInputStream("<tests>", "var test = 10;");
        Lexer lexer = new Lexer(chars);
        TokenInputStream in = lexer.makeTokens();

        assertSame(TokenType.KEYWORD_VAR, in.peek().getType());
        assertSame(null, in.peek().getValue());

        Position start = in.getMap().resolve(in.peek().getStart());
        Position end = in.getMap().resolve(in.peek().getEnd());

        assertSame(0, start.getIndex());
        assertSame(1, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(2, end.getIndex());
        assertSame(3, end.getColumn());
        assertSame(1, end.getLine());

        in.skip();

        assertSame(TokenType.IDENTIFIER, in.peek().getType());
        assertEquals("test", in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(4, start.getIndex());
        assertSame(5, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(7, end.getIndex());
        assertSame(8, end.getColumn());
        assertSame(1, end.getLine());

        in.skip();

        assertSame(TokenType.ASSIGN, in.peek().getType());
        assertSame(null, in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(9, start.getIndex());
        assertSame(10, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(9, end.getIndex());
        assertSame(10, end.getColumn());
        assertSame(1, end.getLine());

        in.skip();

        assertSame(TokenType.INTEGER, in.peek().getType());
        assertEquals("10", in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(11, start.getIndex());
        assertSame(12, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(12, end.getIndex());
        assertSame(13, end.getColumn());
        assertSame(1, end.getLine());

        in.skip();

        assertSame(TokenType.SEMICOLON, in.peek().getType());
        assertSame(null, in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(13, start.getIndex());
        assertSame(14, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(13, end.getIndex());
        assertSame(14, end.getColumn());
        assertSame(1, end.getLine());

    }

    @Test
    public void testGetMultiLinePosition() {

        CharacterInputStream chars = new SourceCharacterInputStream("<tests>", "var test\n  = \n10;");
        Lexer lexer = new Lexer(chars);
        TokenInputStream in = lexer.makeTokens();

        assertSame(TokenType.KEYWORD_VAR, in.peek().getType());
        assertSame(null, in.peek().getValue());

        Position start = in.getMap().resolve(in.peek().getStart());
        Position end = in.getMap().resolve(in.peek().getEnd());

        assertSame(0, start.getIndex());
        assertSame(1, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(2, end.getIndex());
        assertSame(3, end.getColumn());
        assertSame(1, end.getLine());

        in.skip();

        assertSame(TokenType.IDENTIFIER, in.peek().getType());
        assertEquals("test", in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(4, start.getIndex());
        assertSame(5, start.getColumn());
        assertSame(1, start.getLine());

        assertSame(7, end.getIndex());
        assertSame(8, end.getColumn());
        assertSame(1, end.getLine());

        in.skip();

        assertSame(TokenType.LINE_SEPARATOR, in.peek().getType());
        assertSame(null, in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(8, start.getIndex());
        assertSame(0, start.getColumn());
        assertSame(2, start.getLine());

        assertSame(8, end.getIndex());
        assertSame(0, end.getColumn());
        assertSame(2, end.getLine());

        in.skip();

        assertSame(TokenType.ASSIGN, in.peek().getType());
        assertSame(null, in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(11, start.getIndex());
        assertSame(3, start.getColumn());
        assertSame(2, start.getLine());

        assertSame(11, end.getIndex());
        assertSame(3, end.getColumn());
        assertSame(2, end.getLine());

        in.skip();

        assertSame(TokenType.LINE_SEPARATOR, in.peek().getType());
        assertSame(null, in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(13, start.getIndex());
        assertSame(1, start.getColumn());
        assertSame(3, start.getLine());

        assertSame(13, end.getIndex());
        assertSame(1, end.getColumn());
        assertSame(3, end.getLine());

        in.skip();

        assertSame(TokenType.INTEGER, in.peek().getType());
        assertEquals("10", in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(14, start.getIndex());
        assertSame(2, start.getColumn());
        assertSame(3, start.getLine());

        assertSame(15, end.getIndex());
        assertSame(3, end.getColumn());
        assertSame(3, end.getLine());

        in.skip();

        assertSame(TokenType.SEMICOLON, in.peek().getType());
        assertSame(null, in.peek().getValue());

        start = in.getMap().resolve(in.peek().getStart());
        end = in.getMap().resolve(in.peek().getEnd());

        assertSame(16, start.getIndex());
        assertSame(4, start.getColumn());
        assertSame(3, start.getLine());

        assertSame(16, end.getIndex());
        assertSame(4, end.getColumn());
        assertSame(3, end.getLine());

    }

}

File: shake/compiler/lexer/src/commonTest/kotlin/com/shakelang/shake/lexer/TokenPosition.kt
package com.shakelang.shake.lexer

import com.shakelang.shake.lexer.token.ShakeTokenType
import com.shakelang.util.parseutils.characters.streaming.CharacterInputStream
import com.shakelang.util.parseutils.characters.streaming.SourceCharacterInputStream
import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe

@Suppress("unused")
class TokenPosition : FreeSpec(
    {

        "get basic position" {
            val chars: CharacterInputStream = SourceCharacterInputStream("<tests>", "val test = 10;")
            val lexer = ShakeLexer(chars)
            val input = lexer.makeTokens()
            input.peek().type shouldBe ShakeTokenType.KEYWORD_VAL
            input.peek().value shouldBe null
            var start = input.map.resolve(input.peek().start)
            var end = input.map.resolve(input.peek().end)
            start.index shouldBe 0
            start.column shouldBe 1
            start.line shouldBe 1
            end.index shouldBe 2
            end.column shouldBe 3
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.IDENTIFIER
            input.peek().value shouldBe "test"
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 4
            start.column shouldBe 5
            start.line shouldBe 1
            end.index shouldBe 7
            end.column shouldBe 8
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.ASSIGN
            input.peek().value shouldBe null
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 9
            start.column shouldBe 10
            start.line shouldBe 1
            end.index shouldBe 9
            end.column shouldBe 10
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.INTEGER
            input.peek().value shouldBe "10"
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 11
            start.column shouldBe 12
            start.line shouldBe 1
            end.index shouldBe 12
            end.column shouldBe 13
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.SEMICOLON
            input.peek().value shouldBe null
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 13
            start.column shouldBe 14
            start.line shouldBe 1
            end.index shouldBe 13
            end.column shouldBe 14
            end.line shouldBe 1
        }

        "get multi line position" {
            val chars: CharacterInputStream = SourceCharacterInputStream("<tests>", "val test\n  = \n10;")
            val lexer = ShakeLexer(chars)
            val input = lexer.makeTokens()
            input.peek().type shouldBe ShakeTokenType.KEYWORD_VAL
            input.peek().value shouldBe null
            var start = input.map.resolve(input.peek().start)
            var end = input.map.resolve(input.peek().end)
            start.index shouldBe 0
            start.column shouldBe 1
            start.line shouldBe 1
            end.index shouldBe 2
            end.column shouldBe 3
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.IDENTIFIER
            input.peek().value shouldBe "test"
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 4
            start.column shouldBe 5
            start.line shouldBe 1
            end.index shouldBe 7
            end.column shouldBe 8
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.LINE_SEPARATOR
            input.peek().value shouldBe null
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 8
            start.column shouldBe 9
            start.line shouldBe 1
            end.index shouldBe 8
            end.column shouldBe 9
            end.line shouldBe 1
            input.skip()
            input.peek().type shouldBe ShakeTokenType.ASSIGN
            input.peek().value shouldBe null
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 11
            start.column shouldBe 3
            start.line shouldBe 2
            end.index shouldBe 11
            end.column shouldBe 3
            end.line shouldBe 2
            input.skip()
            input.peek().type shouldBe ShakeTokenType.LINE_SEPARATOR
            input.peek().value shouldBe null
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 13
            start.column shouldBe 5
            start.line shouldBe 2
            end.index shouldBe 13
            end.column shouldBe 5
            end.line shouldBe 2
            input.skip()
            input.peek().type shouldBe ShakeTokenType.INTEGER
            input.peek().value shouldBe "10"
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 14
            start.column shouldBe 1
            start.line shouldBe 3
            end.index shouldBe 15
            end.column shouldBe 2
            end.line shouldBe 3
            input.skip()
            input.peek().type shouldBe ShakeTokenType.SEMICOLON
            input.peek().value shouldBe null
            start = input.map.resolve(input.peek().start)
            end = input.map.resolve(input.peek().end)
            start.index shouldBe 16
            start.column shouldBe 3
            start.line shouldBe 3
            end.index shouldBe 16
            end.column shouldBe 3
            end.line shouldBe 3
        }
    },
)
